jose/Jwe.ml

```diff
@@ -22,25 +29,49 @@ Steps to create a JWE
 * Base64url encode the Authentication Tag to create the Encoded JWE Authentication Tag.
 * Assemble the final representation: The Compact Serialization of this result is the concatenation of the Encoded JWE Header, the Encoded JWE Encrypted Key, the Encoded JWE Initialization Vector, the Encoded JWE Ciphertext, and the Encoded JWE Authentication Tag in that order, with the five strings being separated by four period ('.') characters. 
 *)
-(*
-  TODO: Implement encryption
+
 let make_cek enc =
   let key_length = Jwa.enc_to_length enc in
   Mirage_crypto_rng.generate (key_length / 8)
   |> Cstruct.to_string
   |> Jwk.make_oct ~use:`Enc
 
-let encrypt ?protected payload ~(jwk : Jwk.priv Jwk.t) =
-  let () = match protected with Some _ -> () | None -> () in
-  let _input_key =
-    match Jwk.get_alg jwk with
-    | `RSA_OAEP -> Ok (Mirage_crypto_rng.generate 32)
-    | _ -> Error `Unsupported_enc
+let encrypt_payload ~enc ~cek ~init_vector ~aad payload =
+  let iv = Cstruct.of_string init_vector in
+  match enc with
+  | Some `A256GCM ->
+      let cek = Cstruct.of_string cek in
+      let key = Mirage_crypto.Cipher_block.AES.GCM.of_secret cek in
+      let adata = Cstruct.of_string aad in
+      Mirage_crypto.Cipher_block.AES.GCM.encrypt ~key ~iv ~adata
+        (Cstruct.of_string payload)
+      |> fun { message; tag } ->
+      let tag_string = Cstruct.to_string tag in
+      let ciphertext = Cstruct.to_string message in
+      Ok (ciphertext, tag_string)
+  | _ -> Error (`Msg "unsupported encryption")
+
+let encrypt (type a) ~(jwk : a Jwk.t) t =
```

Comment:
what was the decision behind having this function `raise` on failure instead of returning a result type?

Reply:
Just hacking it together currently, will convert to result when it works as part of the cleanup